Source/WebDriver/WebSocketServer.h

```diff
@@ -0,0 +1,152 @@
+    RefPtr<JSON::Object> parameters;
+
+    static std::optional<Command> fromData(const char* data, size_t dataLength);
+};
+
+// https://w3c.github.io/webdriver-bidi/#websocket-listener
+// A WebSocket listener is a network endpoint that is able to accept incoming WebSocket connections.
+struct WebSocketListener : public RefCounted<WebSocketListener> {
+    String host { };
+    unsigned port { 0 };
+    bool isSecure { false };
+    std::vector<String> resources { };
+
+    static RefPtr<WebSocketListener> create(const String& host, unsigned port, bool secure, const std::vector<String>& resources = { })
+    {
+        return adoptRef(*new WebSocketListener(host, port, secure, resources));
+    }
+
+private:
+    WebSocketListener(const String& host, unsigned port, bool isSecure, const std::vector<String>& resources)
+        : host(host)
+        , port(port)
+        , isSecure(isSecure)
+        , resources(resources)
+    {
+    }
+};
+
+
+class WebSocketMessageHandler {
+public:
+
+#if USE(SOUP)
+    typedef GRefPtr<SoupWebsocketConnection> Connection;
+#endif
+
+    struct Message {
+        // Optional connection, as the message might be generated without a connection object available (e.g. inside a method handler).
+        // In this case, it gets associated to the connection when sending the message back to the client.
+        std::optional<Connection> connection;
+        const char* data { nullptr };
+        size_t dataLength { 0 };
+
+        static Message fail(CommandResult::ErrorCode, std::optional<Connection>, std::optional<String> errorMessage = std::nullopt, std::optional<String> commandId = std::nullopt);
+    };
+
+    virtual void handleHandshake(HTTPRequestHandler::Request&&, Function<void(std::optional<HTTPRequestHandler::Response>&& response)>&& replyHandler) = 0;
+    virtual void handleMessage(Message&&, Function<void(Message&&)>&& replyHandler) = 0;
+    virtual void clientDisconnected(Connection) = 0;
+private:
+};
+
+class WebSocketServer : public RefCounted<WebSocketServer>, public CanMakeWeakPtr<WebSocketServer> {
+public:
+    explicit WebSocketServer(WebSocketMessageHandler&, WebDriverService&);
+    virtual ~WebSocketServer() = default;
+
+    std::optional<String> listen(const std::optional<String>& host, unsigned port);
+    bool isListening();
+    void disconnect();
+
+    WebSocketMessageHandler& messageHandler() { return m_messageHandler; }
+
+    const RefPtr<WebSocketListener> listener() const { return m_listener; }
+
+    void addStaticConnection(WebSocketMessageHandler::Connection);
+    bool isStaticConnection(WebSocketMessageHandler::Connection&);
+    void removeStaticConnection(WebSocketMessageHandler::Connection&);
+
+    void addConnection(std::pair<WebSocketMessageHandler::Connection, String>);
+    std::optional<RefPtr<Session>> getSession(WebSocketMessageHandler::Connection&);
+    std::optional<WebSocketMessageHandler::Connection> getConnection(const String& sessionId);
```

Comment:
```suggestion
    WebSocketMessageHandler::Connection& connection(const String& sessionId);
```

Reply:
For this one I kept the optional, as we might not have a connection associated to the given session. Maybe return `Connection` (a `GRefPtr`) by value instead?